### DISCUSSION TITLE
What's the proper way to define a header-only library?

### DISCUSSION THREAD

**Discussion by xq114** [xmake-io/xmake#1725] [Q&A]
```
target("mylibrary")
  set_kind("static") -- 报错找不到lib
  set_kind("phony") -- headerfiles不会被copy
  add_headerfiles("...")
```
怎样设置可以导出header-only的target？

**Comment by waruqi** [comment]
static+置空on_link试试

**Reply by waruqi** [reply]
如果只是 header-only 。。其实直接 phony + on_install 里面自己 os.cp 下 headers 就完事了。。

**Reply by xq114** [reply]
置空on_link实测不行，会报错需要先构建当前target；

phony+on_install可行，但用起来非常麻烦，而且很易错。原本只需要
```
add_headerfiles("armadillo", "(armadillo_bits/*.hpp)")
```
现在要
```
on_install(function (target)
    os.cp("armadillo", path.join(target:installdir(), "include", "armadillo"))
    os.mkdir(path.join(target:installdir(), "include", "armadillo_bits"))
    os.cp("armadillo_bits/*.hpp", path.join(target:installdir(), "include", "armadillo_bits"))
end)
```
注意中间的`os.mkdir`，这是为了使得armadillo_bits不被创建为文件，这里给debug带来困难

如果xmake能在target没有objectfile时自动不链接就好了

**Reply by xq114** [reply]
或者有一个headeronly类型，没有c/cpp文件也能启用public define/package/includedirs和headerfiles功能

**Reply by waruqi** [reply]
不想额外增加基础kind ，新增一个kind，对所有action以及各种case 都会引出更多更复杂的处理

我改进了下 phony，可以处理 add_headerfiles 的 install 。。
